File: backend/app/api/routes/export.py
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.api.deps import require_admin
from app.core.database import get_db
from app.models.user import User
from app.services import export_service

router = APIRouter(prefix="/export", tags=["export"])

_CONTENT_TYPES = {
    "csv": "text/csv",
    "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "pdf": "application/pdf",
}
_BUILDERS = {"csv": export_service.build_csv, "xlsx": export_service.build_xlsx, "pdf": export_service.build_pdf}


@router.get("/{export_format}")
def export_report(export_format: str, db: Session = Depends(get_db), _: User = Depends(require_admin)) -> Response:
    if export_format not in _BUILDERS:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, detail="format must be one of: csv, xlsx, pdf")
    content = _BUILDERS[export_format](db)
    filename = f"hackathon-report.{export_format}"
    return Response(
        content=content,
        media_type=_CONTENT_TYPES[export_format],
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: workers/src/routes/export.ts
import { Hono } from "hono";

import { badRequest } from "../http";
import { requireAdmin, requireUser, type AppEnv } from "../middleware/auth";
import { buildCsv, buildPdf, buildXlsx } from "../services/export";

const CONTENT_TYPES: Record<string, string> = {
  csv: "text/csv",
  xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  pdf: "application/pdf",
};

const BUILDERS: Record<string, (db: D1Database, eventsDb: D1Database) => Promise<Uint8Array>> = {
  csv: buildCsv,
  xlsx: buildXlsx,
  pdf: buildPdf,
};

export const exportRoutes = new Hono<AppEnv>();

exportRoutes.use("*", requireUser, requireAdmin);

exportRoutes.get("/:export_format", async (c) => {
  const format = c.req.param("export_format");
  const builder = BUILDERS[format];
  if (!builder) throw badRequest("format must be one of: csv, xlsx, pdf");

  const content = await builder(c.env.DB, c.env.EVENTS_DB);
  return new Response(content, {
    headers: {
      "Content-Type": CONTENT_TYPES[format]!,
      "Content-Disposition": `attachment; filename="hackathon-report.${format}"`,
    },
  });
});
